CsMerge/MergeHelper.cs

```diff
@@ -17,52 +17,65 @@ public static class MergeHelper<T> where T: class, IKeyedEntry {
 
+      string key = ( b ?? m ?? t ).Key;
+
       if ( Equals( b, t ) && Equals( b, m ) ) {
-        return new MergeResult<T>( b, MergeType.NoChanges );
+        return new MergeResult<T>( key, b, MergeType.NoChanges );
       }
 
       if ( b == null ) {
         if ( Equals( m, t ) ) {
-          return new MergeResult<T>( m, MergeType.LocalAdded | MergeType.IncomingAdded );
+          return new MergeResult<T>( key, m, MergeType.LocalAdded | MergeType.IncomingAdded );
         }
+
         if ( t != null && m != null ) {
-          return new MergeResult<T>( contentResolver( new Conflict<T>( b, m, t ) ),
-            MergeType.LocalAdded | MergeType.IncomingAdded );
+          return new MergeResult<T>( key, contentResolver( new Conflict<T>( b, m, t ) ), MergeType.LocalAdded | MergeType.IncomingAdded );
         }
-        return new MergeResult<T>( t ?? m, t == null ? MergeType.LocalAdded : MergeType.IncomingAdded );
+
+        return new MergeResult<T>( key, t ?? m, t == null ? MergeType.LocalAdded : MergeType.IncomingAdded );
       }
 
       if ( m == null && t == null ) {
-        return new MergeResult<T>( MergeType.LocalDeleted | MergeType.IncomingDeleted );
+        return new MergeResult<T>( key, MergeType.LocalDeleted | MergeType.IncomingDeleted );
       }
 
       if ( m == null ) {
         if ( t.Equals( b ) ) {
-          return new MergeResult<T>( MergeType.LocalDeleted );
+          return new MergeResult<T>( key, MergeType.LocalDeleted );
         }
 
         // Local deleted something modified in incoming
         var resolved = modDeleteResolver( new Conflict<T>( b, m, t ) );
-        return new MergeResult<T>( resolved, MergeType.LocalDeleted | MergeType.IncomingModified );
+        return new MergeResult<T>( key, resolved, MergeType.LocalDeleted | MergeType.IncomingModified );
       }
 
       if ( t == null ) {
         if ( Equals( m, b ) ) {
-          return new MergeResult<T>( MergeType.IncomingDeleted );
+          return new MergeResult<T>( b.Key, MergeType.IncomingDeleted );
         }
 
         // Incoming deleted something modified in local
         var resolved = modDeleteResolver( new Conflict<T>( b, m, t ) );
-        return new MergeResult<T>( resolved, MergeType.LocalModified | MergeType.IncomingDeleted );
+        return new MergeResult<T>( key, resolved, MergeType.LocalModified | MergeType.IncomingDeleted );
       }
-
+      
       if ( Equals( m, t ) ) {
-        return new MergeResult<T>( m, MergeType.LocalModified | MergeType.IncomingModified );
+        return new MergeResult<T>( key, m, MergeType.LocalModified | MergeType.IncomingModified );
       }
 
-      return new MergeResult<T>(
+      if ( Equals( b, m ) && !Equals( b, t ) ) {
+        return new MergeResult<T>( key, t, MergeType.IncomingModified );
+      }
+
+      if ( Equals( b, t ) && !Equals( b, m ) ) {
+        return new MergeResult<T>( key, m, MergeType.LocalModified );
+      }
```

Comment:
I've added a new auto merge condition - if a reference is unchanged in theirs, but changed in mine, we can auto-resolve to mine.